### DISCUSSION TITLE
Empty tcrp, can't boot

### DISCUSSION THREAD

**Discussion by f22raptorroland** [PeterSuh-Q3/tinycore-redpill#33] [Q&A] [answered]
Thanks for the awesome work!
#### One week in haven
I managed to install it last week, and was up and running with no Problem. After a power loss when I rebooted it wouldn't boot anymore. 

I tried to re-build, re-install the image, different flash drives, etc, but no luck. 

#### Setup
- I'm on bare metal
- flash drive connected another soldered on the Motherboard(unused 1GB)
- Terminal from motherboard

#### Output
<details>
  <summary><b>Output on boot</B></summary>

   `Loading Linux...
vga=791 is deprecated. Use set gfxpayload=1024x768x16,1024x768 before linux
command instead.
Loading initramfs...
Booting TinyCore Friend
Saving 256 bits of non-creditable seed for next boot
Starting syslogd: OK
Starting acpid: OK
Starting klogd: OK
Running sysctl: OK
Populating /dev using udev: done
Starting system message bus: done
Starting iptables: OK
Starting network: OK
Starting dhcpcd...
dhcpcd-10.0.4 starting
dev: loaded udev
no interfaces have a carrier
Starting connmand: OK
ssh-keygen: generating new host keys: RSA ECDSA ED25519
Starting sshd: OK
Starting SMB services: OK
Starting NMB services: OK

Please press Enter to activate this console. [2024/03/20 20:55:03:0670] N: ttyd 1.7.3 (libwebsockets 4.3.2-2023.08.4)
[2024/03/20 20:55:03:0672] N: tty configuration:
[2024/03/20 20:55:03:0673] N:   start command: login -f root
[2024/03/20 20:55:03:0673] N:   close signal: SIGHUP (1)
[2024/03/20 20:55:03:0673] N:   terminal type: xterm-256color
[2024/03/20 20:55:03:0686] N:    /usr/lib/libwebsockets-evlib_uv.so
[2024/03/20 20:55:03:0687] N: lws_create_context: LWS: 4.3.2-2023.08.4, NET CLI SRV H1 H2 WS ConMon IPV6-off
[2024/03/20 20:55:03:0688] N: elops_init_pt_uv:  Using foreign event loop...
[2024/03/20 20:55:03:0688] N: __lws_lc_tag:  ++ [wsi|0|pipe] (1)
[2024/03/20 20:55:03:0689] N: __lws_lc_tag:  ++ [vh|0|netlink] (1)
[2024/03/20 20:55:03:0690] N: __lws_lc_tag:  ++ [vh|1|default||7681] (2)
[2024/03/20 20:55:03:0691] N: [vh|1|default||7681]: lws_socket_bind: source ads 0.0.0.0
[2024/03/20 20:55:03:0691] N: __lws_lc_tag:  ++ [wsi|1|listen|default||7681] (2)
[2024/03/20 20:55:03:0692] N:  Listening on port: 7681`

</details>

<details>
  <summary><b>After boot pressing key / running boot.sh</B></summary>

   `root@tcrpfriend:~# ./boot.sh
device node not found
model is not resolved. Please check the /mnt/tcrp/user_config.json file. stopping boot process

!!NOTE: This file is empty!!
root@tcrpfriend:# cat  /mnt/tcrp/user_config.json
root@tcrpfriend:#`

</details>

#### Populating user_config.json

At this point I can see the it on the network at  \\192.168.1.198\tcrpfriend\tcrp
There is an empty user_config.json
I added the same config used at build time.
But both tcrp-p1, and tcrp-p2 are empty observing from the network 

<details>
  <summary><b>Output after config</B></summary>

   `root@tcrpfriend:# ./boot.sh                                                    device node not found
User config file needs update, updating -> Done
cat: /mnt/tcrp-p1/GRUB_VER: No such file or directory
--------------------------------------={ TinyCore RedPill Friend }=--------------------------------------
TCRP Friend Version : 0.1.0t
0.1.0  friend kernel version up from 5.15.26 to 6.4.16
0.1.0l Modified the kexec option from -a (memory) to -f (file) to accurately load the patched initrd-dsm.
0.1.0m Recycle initrd-dsm instead of custom.gz (extract /exts)
0.1.0n When a loader is inserted into syno disk /dev/sda and /dev/sdb,
       change to additionally mount partitions 1,2 and 3 to /dev/sda5,/dev/sda6 and /dev/sdb5.
0.1.0o Added RedPill bootloader hard disk porting function
0.1.0p Added priority search for USB or VMDK bootloader over bootloader injected into HDD
0.1.0q Added support for SHR type to HDD for bootloader injection.
       synoboot3 unified to use partition number 4 instead of partition number 5 (1 BASIC + 1 SHR required)
0.1.0r Fix bug of 0.1.0q (Fix typo for partition number 4)
0.1.0s Force the dom_szmax limit of the injected bootloader to be 16GB
0.1.0t Supports bootloader injection with SHR disk only
       dom_szmax=32GB (limit size of the injected bootloader)

Model : DS3622xs+, Serial : 20C0SQRTZSWZA, Mac :c1d297c8b765, Build : 7.2.1-69057, Update : 0, LKM : prod
Loader BUS:
CPU,MEM:  Intel(R) Atom(TM) CPU D525   @ 1.80GHz [4 Thread(s)], 1.9G Memory
DMI: Iomega Corporation StorCenter Pro xxxx/StorCenter Pro xxxx, BIOS 0122 03/26/2013
SAS/RAID HBAs Count : 0 , NICs Count : 2
System is running in Legacy boot mode
Setting MAC Address from c1:d2:97:c8:b7:65 to c1:d2:97:c8:b7:65 on eth0 (r8169)
Setting MAC Address from c1:d2:97:c8:b7:66 to c1:d2:97:c8:b7:66 on eth1 (r8169)
[2024/03/20 20:59:45:5910] N: rops_handle_POLLIN_netlink: DELADDR
IP Address : 192.168.1.198, Network Interface Card : eth0 [10ec:8168] (r8168
r8169)
IP Address : , Network Interface Card : eth1 [10ec:8168] (r8168
r8169)
ERROR ! /mnt/tcrp-p2/rd.gz file not found, stopping boot process
root@tcrpfriend:#`
</details>

#### Honorable mentions in log

<details>
  <summary><i>When running Re-Install DSM</i></summary>

  :: Mounting devtmpfs ... [FAILED]
:: Loading module acpi-cpufreq ... [FAILED]
:: Loading module i40e ... [FAILED]

[    7.221787] redpill: module verification failed: signature and/or required key missing - tainting kernel

`
Checking ext4 rootfs on /dev/md0
return value: 1
All components of md0 have partition layout above v9.
ForceNewestLayout: Skipped
[  114.854048] EXT4-fs (md0): couldn't mount as ext3 due to feature incompatibilities
[  114.916348] EXT4-fs (md0): mounted filesystem with ordered data mode. Opts: (null)
Mounting /dev/md0 /tmpRoot
[  115.145681] EXT4-fs (m[user_config.json](https://github.com/PeterSuh-Q3/tinycore-redpill/files/14672976/user_config.json)): couldn't mount as ext3 due to feature incompatibilities
[  115.163855] EXT4-fs (md0): mounted filesystem with ordered data mode. Opts: prjquota,rootprjquota,barrier=1
insmod: can't insert '/usr/lib/modu[  115.209979] eth0: 0xffffc900000ee000, c1:d2:97:c8:b7:6b, IRQ 29
les/r8168.ko': File exists
[  115.426200] Module [r8168] is removed.
Update OOB FW A success!

</details>

##### Constant error

<details>
  <summary><i>Some constant warning</i></summary>
I get this every minute or so, but I got it even when the thing was working Maybe from the onboard USB storage

[  202.162026] usb 8-1: device descriptor read/64, error -110
[  217.394029] usb 8-1: device descriptor read/64, error -110
[  217.623029] usb 8-1: new full-speed USB device number 3 using uhci_hcd

</details>

#### Things I tried

- Re-Install DSM of ... to USB and SATA (I'm not sure what is the difference)
- Different Flash Drive
- I had 4 HDDs in RAID 5 I unplugged the others formatted the remaining(from menu.sh), and tried to install it on that(I don't have any more spared)
- Rebuilding with different parameters

### Bonus question

Now that I have formatted one of the 4 drives from Raid 5 it should theoretically still be possible to recover the data if I get it working right? Is it complicated? 

I've only had a week's worth of data there, but I recovered some old pictures from a failed drive and I'm not sure if it survives an other deep scan

Could someone help me with some guidance, please? 
Where should I look for the problem?

[user_config.json](https://github.com/PeterSuh-Q3/tinycore-redpill/files/14672980/user_config.json)
[When booting Re-install.log](https://github.com/PeterSuh-Q3/tinycore-redpill/files/14672981/When.booting.Re-install.log)

**Comment by PeterSuh-Q3** [comment]
It has been moved from the issues section to the discussion section.
This is because it is easy to translate into Korean. Please understand.

The boot device you soldered in must be either Sata or USB.
However, it seems that the device cannot be found in the log results.
Loader BUS:
ATA or USB should be output.
Among the disks under /dev/sd#
/dev/sdv appears to be a loader device.
Please let me know the results of executing the command below from the friend prompt that stopped due to an error.


fdisk -l | grep "Disk /dev/sd" 

udevadm info --query property --name /dev/sdv

**Comment by f22raptorroland** [comment]
Please excuse me, I didn't know. 
If you prefer I can also translate my response into Korean instead.

Thank you for getting back to me so quickly. 

### Results

Note: (I'm not sure if it is important)
My boot device is not directly soldered in. It is connected to a USB port. 
I have another soldered-in flash device. (That has the unused factory bootloader 1GB I think)

#### Outputs
<details>
  <summary><i>fdisk -l | grep "Disk /dev/sd</i></summary>

root@tcrpfriend:# "
Disk /dev/sda: 3.64 TiB, 4000787030016 bytes, 7814037168 sectors
Disk /dev/sdb: 956 MiB, 1002438656 bytes, 1957888 sectors
Disk /dev/sdc: 57.73 GiB, 61991813632 bytes, 121077761 sectors

It is a 64Gb Kingstone Datatravel

</details>


**There is no SDV**, but there is a,b,c
<details>
  <summary><i>udevadm info --query property --name /dev/sdv</i></summary>

device node not found

</details>



<details>
  <summary><i>udevadm info --query property --name /dev/sda</i></summary>

DEVLINKS=/dev/disk/by-diskseq/9 /dev/disk/by-id/ata-ST4000VN008-2DR166_ZDH9KSDT /dev/disk/by-id/wwn-0x5000c500c92b036f
DEVNAME=/dev/sda
DEVPATH=/devices/pci0000:00/0000:00:1f.2/ata1/host0/target0:0:0/0:0:0:0/block/sda
DEVTYPE=disk
DISKSEQ=9
ID_ATA=1
ID_ATA_DOWNLOAD_MICROCODE=1
ID_ATA_FEATURE_SET_APM=1
ID_ATA_FEATURE_SET_APM_CURRENT_VALUE=128
ID_ATA_FEATURE_SET_APM_ENABLED=1
ID_ATA_FEATURE_SET_HPA=1
ID_ATA_FEATURE_SET_HPA_ENABLED=1
ID_ATA_FEATURE_SET_PM=1
ID_ATA_FEATURE_SET_PM_ENABLED=1
ID_ATA_FEATURE_SET_PUIS=1
ID_ATA_FEATURE_SET_PUIS_ENABLED=0
ID_ATA_FEATURE_SET_SECURITY=1
ID_ATA_FEATURE_SET_SECURITY_ENABLED=0
ID_ATA_FEATURE_SET_SECURITY_ENHANCED_ERASE_UNIT_MIN=430
ID_ATA_FEATURE_SET_SECURITY_ERASE_UNIT_MIN=430
ID_ATA_FEATURE_SET_SMART=1
ID_ATA_FEATURE_SET_SMART_ENABLED=0
ID_ATA_ROTATION_RATE_RPM=5980
ID_ATA_SATA=1
ID_ATA_SATA_SIGNAL_RATE_GEN1=1
ID_ATA_SATA_SIGNAL_RATE_GEN2=1
ID_ATA_WRITE_CACHE=1
ID_ATA_WRITE_CACHE_ENABLED=1
ID_BUS=ata
ID_MODEL=ST4000VN008-2DR166
ID_MODEL_ENC=ST4000VN008-2DR166\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20
ID_PART_TABLE_TYPE=gpt
ID_PART_TABLE_UUID=b7269dc4-e473-435b-ac0d-530937d16ecb
ID_REVISION=SC60
ID_SERIAL=ST4000VN008-2DR166_ZDH9KSDT
ID_SERIAL_SHORT=ZDH9KSDT
ID_TYPE=disk
ID_WWN=0x5000c500c92b036f
ID_WWN_WITH_EXTENSION=0x5000c500c92b036f
MAJOR=8
MINOR=0
SUBSYSTEM=block
USEC_INITIALIZED=20463773
root@tcrpfriend:#

</details>







<details>
  <summary><i>udevadm info --query property --name /dev/sdb</i></summary>

DEVLINKS=/dev/disk/by-diskseq/10 /dev/disk/by-id/usb-SMI_USB_DISK_AA04012700008612-0:0 /dev/disk/by-path/pci-0000:00:1d.7-usb-0:1:1.0-scsi-0:0:0:0
DEVNAME=/dev/sdb
DEVPATH=/devices/pci0000:00/0000:00:1d.7/usb3/3-1/3-1:1.0/host6/target6:0:0/6:0:0:0/block/sdb
DEVTYPE=disk
DISKSEQ=10
ID_BUS=usb
ID_INSTANCE=0:0
ID_MODEL=USB_DISK
ID_MODEL_ENC=USB\x20DISK\x20\x20\x20\x20\x20\x20\x20\x20
ID_MODEL_ID=1000
ID_PART_TABLE_TYPE=dos
ID_PART_TABLE_UUID=000748a4
ID_PATH=pci-0000:00:1d.7-usb-0:1:1.0-scsi-0:0:0:0
ID_PATH_TAG=pci-0000_00_1d_7-usb-0_1_1_0-scsi-0_0_0_0
ID_REVISION=0100
ID_SERIAL=SMI_USB_DISK_AA04012700008612-0:0
ID_SERIAL_SHORT=AA04012700008612
ID_TYPE=disk
ID_USB_DRIVER=usb-storage
ID_USB_INTERFACES=:080650:
ID_USB_INTERFACE_NUM=00
ID_VENDOR=SMI
ID_VENDOR_ENC=SMI\x20\x20\x20\x20\x20
ID_VENDOR_ID=090c
MAJOR=8
MINOR=16
SUBSYSTEM=block
USEC_INITIALIZED=20423364
root@tcrpfriend:#  

  </details>





<details>
  <summary><i>udevadm info --query property --name /dev/sdc</i></summary>

DEVLINKS=/dev/disk/by-diskseq/11 /dev/disk/by-id/usb-Kingston_DataTraveler_3.0_E0D55E6C71161580E9118F56-0:0 /dev/disk/by-path/pci-0000:00:1d.7-usb-0:4:1.0-scsi-0:0:0:0
DEVNAME=/dev/sdc
DEVPATH=/devices/pci0000:00/0000:00:1d.7/usb3/3-4/3-4:1.0/host7/target7:0:0/7:0:0:0/block/sdc
DEVTYPE=disk
DISKSEQ=11
ID_BUS=usb
ID_INSTANCE=0:0
ID_MODEL=DataTraveler_3.0
ID_MODEL_ENC=DataTraveler\x203.0
ID_MODEL_ID=1666
ID_PART_TABLE_TYPE=dos
ID_PART_TABLE_UUID=f110ee87
ID_PATH=pci-0000:00:1d.7-usb-0:4:1.0-scsi-0:0:0:0
ID_PATH_TAG=pci-0000_00_1d_7-usb-0_4_1_0-scsi-0_0_0_0
ID_REVISION=0000
ID_SERIAL=Kingston_DataTraveler_3.0_E0D55E6C71161580E9118F56-0:0
ID_SERIAL_SHORT=E0D55E6C71161580E9118F56
ID_TYPE=disk
ID_USB_DRIVER=usb-storage
ID_USB_INTERFACES=:080650:
ID_USB_INTERFACE_NUM=00
ID_VENDOR=Kingston
ID_VENDOR_ENC=Kingston
ID_VENDOR_ID=0951
MAJOR=8
MINOR=32
SUBSYSTEM=block
USEC_INITIALIZED=20445463
root@tcrpfriend:#

</details>


#### Sum

**So basically**
- sda: SATA disc I wiped earlier
- sdb: Unused 1GB soldered "USB" flash
- sdc: The Pendrive plugged in with tinycore


### Notes

Also, I see this on the network. A shared drive appears with three folders:
![image](https://github.com/PeterSuh-Q3/tinycore-redpill/assets/3228649/9b9670ef-ae23-4959-875b-038b17d4d6df)

This is the structure:
![image](https://github.com/PeterSuh-Q3/tinycore-redpill/assets/3228649/b7272f57-5e5b-4fdc-a192-5e2e6fccabe4)


#### Ghost sdv
I don't know why is there no SDV suddenly.


<details>
  <summary>boot log of reinstall (When booting Re-install.log)</summary>
<pre>
========== BEGIN DUMP OF ALL PARTITIONS DETECTED ===========
/dev/sda1                  8192        16785407        16777216  fd
/dev/sda2              16785408        20979711         4194304  fd
/dev/sda3              21241856      7813832351      7792590496  fd
</pre>
<pre>
/dev/sdb1                  8192        16785407        16777216  fd
/dev/sdb2              16785408        20979711         4194304  fd
/dev/sdb3              21241856      7813832351      7792590496  fd
</pre>
<pre>
/dev/sdc1                  8192        16785407        16777216  fd
/dev/sdc2              16785408        20979711         4194304  fd
/dev/sdc3              21241856      5860328351      5839086496  fd
</pre>
<pre>
/dev/sdd1                  8192        16785407        16777216  fd
/dev/sdd2              16785408        20979711         4194304  fd
/dev/sdd3              21241856      5860328351      5839086496  fd
</pre>
<pre>
/dev/sdv1                  2048          147455          145408  83
/dev/sdv2                147456          301055          153600  83
/dev/sdv3                301056         3147775         2846720  83
</pre>
  </details>

**Comment by PeterSuh-Q3** [comment]
Do you suspect that /dev/sdb was a tcrp bootloader used in the past and was damaged due to a power outage?
/dev/sdc appears to be the newly added USB pen drive.

I think you need to check additionally with the two commands below to be sure.
Please perform this at the friend prompt.

fdisk -l /dev/sdb
fdisk -l /dev/sdc

TCRP requires only one bootloader.
In particular, TCRP-mshell takes the device it finds first when there are two or more bootloaders.

I think the problematic /dev/sdb is being used first.
If you can remove /dev/sdb, proceed with the removal.
If removal is impossible, it is best to reset it.
If you need to format after initialization, it is recommended to have only one partition with ext partition type rather than vfat.

/dev/sdb, /dev/sdc
Since both are USB-type devices, they may affect bootloader booting.

**Comment by f22raptorroland** [comment]
> Do you suspect that /dev/sdb was a tcrp bootloader used in the past and was damaged due to a power outage?

No. 
I mean dev/sdb has the original bootloader from the manufacturer. It is iomega px4-300, and that has the original image, but it is only 1GB, so I am not using it. (Also it is kind of a fallback solution until I get familiar with the tinycore I wouldn't want to delete it)
I disabled it in the Bios boot menu, but it seems it only hides it temporarily for boot. ( it still shows up for fdisk)



> In particular, TCRP-mshell takes the device it finds first when there are two or more bootloaders.

But it was always like that right? 
Because I originally set up the NAS on the 13th a week ago, and then it was working fine. 
Can it be that it stopped working due to some changes in the past weeks?

#### USB prioiority
I've also filled out VID/PID in the user config.json 
Shouldn't it find the correct USB device based on that?

### Outputs
<details>
  <summary><i>fdisk -l /dev/sdb</i></summary>

Disk /dev/sdb: 956 MiB, 1002438656 bytes, 1957888 sectors
Disk model: USB DISK
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x000748a4

Device     Boot Start     End Sectors  Size Id Type
/dev/sdb1        2048 1957887 1955840  955M 83 Linux
tc@box:$


It is a 64Gb Kingstone Datatravel

</details>


<details>
  <summary><i>fdisk -l /dev/sdc</i></summary>

Disk /dev/sdc: 57.73 GiB, 61991813632 bytes, 121077761 sectors
Disk model: DataTraveler 3.0
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0xf110ee87

Device     Boot  Start     End Sectors  Size Id Type
/dev/sdc1  *      2048  147455  145408   71M 83 Linux
/dev/sdc2       147456  301055  153600   75M 83 Linux
/dev/sdc3       301056 3147775 2846720  1.4G 83 Linux
tc@box:~$

It is a 64Gb Kingstone Datatravel
</details>

I am a bit reluctant to format the dev/sdb flash drive, because that has the original software, and now that is the only thing that is working.
My original plans were to try tinycore, and if it works take apart the device, and re-solder the 1GB memory module to something bigger, and put tinycore there instead
So I would like to save the data there for now. Is it possible to unmount, or hide it somehow from the system during, or right after boot? 


### Update
I also used to give **device node not found**, and I see the effect of yesterday's changes, so instead now, I get **Not Supported Loader BUS Type, program Exit!!!**

Since then I don't see the mounted folders if I access the drive through network sharing(I mean tcrp, tcrp-p1, p2 are not visible anymore)

**Reply by PeterSuh-Q3** [reply, marked as answer]
In that case, you should leave /dev/sdb and skip it.
I am planning to modify and distribute the function to find the bootloader in friends.
I think it will be friend version v0.1.0v.

Recent changes add the ability to use nvme and mmc as loader boot devices.
In the process, boot devices that do not correspond to usb / sata / nvme / mmc, etc. may be recognized as errors.
Since this is the case for your 1G USB DISK device, we are adding a function to skip it.

I'm not sure if you can understand the script, but change it like below.
https://github.com/PeterSuh-Q3/tcrpfriend/commit/ab0468a9d018abfcd225590e788afe0cecdfc5a4
I will distribute it after finishing testing.

**Reply by f22raptorroland** [reply]
So cool! 
Thank you so much for the update.

I am an embedded software engineer, so I know scripts, just not overly familiar with Unix systems

It seems so that it worked with the diff you sent. 
It still has some error when it tries to install DSM, but I'll try, and re-install from scratch to make sure.

Will keep you posted
I'm so happy!!

**Reply by PeterSuh-Q3** [reply]
Even if /dev/sdb exists, menu_m.sh, which was changed to the latest version, does not seem to be affected.
is not it?

The final changes to menu_m.sh are as follows:

https://github.com/PeterSuh-Q3/tinycore-redpill/commit/0b9f22e85dee979bcdbe60e0f6735f5091aceb17

update_tinycore() has nothing to do with updating menu_m.sh.
It only serves to change the tinycore Linux kernel version from 12.0 to 14.0.

Can you take a screenshot of what the latest menu_m.sh looks like on your PC?

Separately from this, friend v0.1.0v has now been released.
If friend booting is possible, try friend booting first rather than trying to enter menu_m.
To automatically update to v0.1.0v on a stopped friend kernel, run the command as shown below.

./down.sh
./boot.sh

**Reply by f22raptorroland** [reply]
Thank you so much for the continuous help, and suggestions.
The diff of boot.sh really did solve the issue. 

After reinstalling while migrating from the three original drives I was able to start up, and begin a repair of my storage pool.
![image](https://github.com/PeterSuh-Q3/tinycore-redpill/assets/3228649/52b24208-142c-4635-9eb0-f2919230def1)

I am extremely happy

**Reply by PeterSuh-Q3** [reply]
Thank you for your donation.

**Reply by f22raptorroland** [reply]
Thank you for the amazing work, and support!
